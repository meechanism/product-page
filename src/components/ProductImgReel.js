import React from 'react';
import styled from 'styled-components';

import Colors from '../styles/colors';

const ImgReelWrapper = styled.div`
  margin: 0;
  display: flex;
  justify-content: center;
`;

const Img = styled.img`
  width: 100%;
  max-height: 80px;
  border-radius: 2px;
`;

export const ImgWrapper = styled.a`
  display: block;
  width: 95px;
  padding: 0.5rem;
  &:hover {
    cursor: pointer;
    opacity: 0.75;
  }
  border-radius: 2px;
  border: 1px solid ${(props) => (props.active ? Colors.gray : 'transparent')};
`;

export const saveImg = (save, index) => (e) => {
  e && e.preventDefault();
  save(index);
};

const ProductImgReel = ({ images, setCurrImgIndex, currImgIndex }) => (
  <ImgReelWrapper>
    {images.map((currImg, i) => {
      const key = `product-image-${i}`;
      return (
        <ImgWrapper
          href=""
          key={key}
          onClick={saveImg(setCurrImgIndex, i)}
          active={currImgIndex === i}>
          <Img src={currImg.href} alt={key} />
        </ImgWrapper>
      );
    })}
  </ImgReelWrapper>
);

export default ProductImgReel;
