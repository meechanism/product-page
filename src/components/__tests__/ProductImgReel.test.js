import React from 'react';
import ProductImgReel, { ImgWrapper, saveImg } from '../ProductImgReel';
import productData from '../../data/products';

describe('ProductImgReel component', () => {
  const testProduct = productData.groups[0];
  const setImgFn = jest.fn();
  const { images } = testProduct;
  const e = { preventDefault: jest.fn() };

  it('should render', () => {
    const output = shallow(
      <ProductImgReel
        images={images}
        currImgIndex={0}
        setCurrImgIndex={setImgFn}
      />
    );
    expect(shallowToJson(output)).toMatchSnapshot();
  });

  it('should save index of img', () => {
    const index = 2;
    saveImg(setImgFn, index)(e);
    expect(setImgFn).toHaveBeenCalledWith(index);
    expect(e.preventDefault).toHaveBeenCalled();
  });

  it('should render thumbnails for all of the product images', () => {
    const output = shallow(
      <ProductImgReel
        images={images}
        currImgIndex={0}
        setCurrImgIndex={setImgFn}
      />
    );
    expect(output.find(ImgWrapper).length).toBe(images.length);
  });

  it('should trigger callback when image is clicked ', () => {
    const output = shallow(
      <ProductImgReel
        images={images}
        currImgIndex={0}
        setCurrImgIndex={setImgFn}
      />
    );
    const firstImg = output.find(ImgWrapper).first();
    firstImg.simulate('click');
    expect(setImgFn).toHaveBeenCalledWith(0);
  });

  it('should set active state on img with correct active index', () => {
    const activeIndex = 1;
    const output = shallow(
      <ProductImgReel
        images={images}
        currImgIndex={activeIndex}
        setCurrImgIndex={setImgFn}
      />
    );
    output.find(ImgWrapper).forEach((img, i) => {
      if (activeIndex !== i) expect(img.props().active).toBe(false);
      else expect(img.props().active).toBe(true);
    });
  });
});
