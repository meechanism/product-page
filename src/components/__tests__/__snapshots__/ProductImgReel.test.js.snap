// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImgReel component should render 1`] = `
<styled.div>
  <styled.a
    active={true}
    href=""
    key="product-image-0"
    onClick={[Function]}
  >
    <styled.img
      alt="product-image-0"
      src="https://www.westelm.com/weimgs/ab/images/wcm/products/202003/0005/belgian-linen-ladder-stripe-embroidery-duvet-cover-shams-6-m.jpg"
    />
  </styled.a>
  <styled.a
    active={false}
    href=""
    key="product-image-1"
    onClick={[Function]}
  >
    <styled.img
      alt="product-image-1"
      src="https://www.westelm.com/weimgs/ab/images/wcm/products/202003/0012/belgian-linen-ladder-stripe-embroidery-duvet-cover-shams-m.jpg"
    />
  </styled.a>
</styled.div>
`;
